### ISSUE TITLE
Reproducible? Same seed and config, but different results in two runs.

### ISSUE THREAD

**Issue by fx-hit** [microsoft/CogACT#37]
The code in `prismatic/vla/datasets/rlds/dataset.py` involves TensorFlow operations for reading data:  

```python
import dlimp as dl

# Interleave at the Frame Level  
dataset: dl.DLataset = dl.DLataset.sample_from_datasets(datasets, sample_weights)  

# Validation =>> fix a single shuffle buffer of data and cache it in RAM; prevents gradual memory increase!  
if not train:  
    dataset = dataset.take(shuffle_buffer_size).cache()  

# Shuffle the Dataset  
#   =>> IMPORTANT :: Shuffle AFTER .cache(), or else memory will still leak!  
dataset = dataset.shuffle(shuffle_buffer_size)  
```  

The original code might have overlooked setting the random seed for this part, as `prismatic/util/torch_utils.py` only sets seeds for `torch`, `random`, and `np.random`:

```python
def set_global_seed(seed: int, get_worker_init_fn: bool = False) -> Optional[Callable[[int], None]]:
    """Sets seed for all randomness libraries (mostly random, numpy, torch) and produces a `worker_init_fn`"""
    assert np.iinfo(np.uint32).min < seed < np.iinfo(np.uint32).max, "Seed outside the np.uint32 bounds!"

    # Set Seed as an Environment Variable
    os.environ["EXPERIMENT_GLOBAL_SEED"] = str(seed)
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)

    return worker_init_function if get_worker_init_fn else None
```

I added `tf.random.set_seed(seed)`, but the training loss still varies between two runs.  

Have you encountered these issues before? How should the random seeds be configured to ensure reproducibility?

**Comment by fx-hit** [comment]
I noticed that TFDS here processes the chunked data files in parallel, causing the order in which they arrive in the buffer to be inconsistent. The shuffle operation is performed on the data within the buffer. Even though the random seed is controlled to be consistent, and the selected indices (idx) may be the same across two runs, the data within the buffer differs, ultimately leading to inconsistencies in the order of data read. This affects the PyTorch training code—running the same configuration twice results in inconsistent data reading, causing different training losses and making reproduction impossible.

**Comment by arnoldland** [comment]
Yes, your observation is correct. This issue arises from how TFDS handles data processing. We previously modified the dlimp library and properly fixed the random seed within the dataloader. In this way, the data placed into the shuffle buffer were exactly the same each time and could be resumed after interruption. However, fixing the seed reduced the performance of RLDS and made training more prone to interruptions (for the reason, you can refer to their documentation for more detailed explanations). After careful consideration and testing, we decided to keep the version with unfixed seeds for RLDS.

If you're interested in exploring how to fix the random seed in this part, you can refer to our modified version of dlimp:  https://github.com/Eureka6174/dlimp

**Comment by fx-hit** [comment]
Thank you for your reply. We have resolved this issue but encountered a new non-reproducible problem.  

We intercepted a small-scale dataset from the tfds data reading pipeline and stored it locally, then handed it over to PyTorch for reading. In other words, the code has been completely switched to PyTorch and no longer involves tfds data reading.  

We have confirmed that all random seeds are fixed, and the data reading order is strictly controlled to be identical. However, when running two experiments with the same configuration, the training loss still varies. This also affects the final success rate evaluation, causing a discrepancy of 5%-10% when evaluating the success rate on window X.  

We found that if we enable autocast for the forward pass in fp16, the training loss becomes consistent across two runs—but the model produces NaN after a few hundred iterations, requiring GradScaler. However, once GradScaler is added, the loss inconsistency reappears.  
Whether the forward autocast is set to the default bf16 or manually switched to fp32, the training loss remains inconsistent between two experiments.  
If we freeze all parameters of the VLM, the training loss becomes consistent across two runs.  
The reason may lie in the backpropagation of DIT's loss to the shallow layers of the VLM, where the gradients are too small, leading to some kind of numerical instability?  
This issue currently has no solution.

**Comment by LukeLIN-web** [comment]
> Thank you for your reply. We have resolved this issue but encountered a new non-reproducible problem.
> 
> We intercepted a small-scale dataset from the tfds data reading pipeline and stored it locally, then handed it over to PyTorch for reading. In other words, the code has been completely switched to PyTorch and no longer involves tfds data reading.
> 
> We have confirmed that all random seeds are fixed, and the data reading order is strictly controlled to be identical. However, when running two experiments with the same configuration, the training loss still varies. This also affects the final success rate evaluation, causing a discrepancy of 5%-10% when evaluating the success rate on window X.
> 
> We found that if we enable autocast for the forward pass in fp16, the training loss becomes consistent across two runs—but the model produces NaN after a few hundred iterations, requiring GradScaler. However, once GradScaler is added, the loss inconsistency reappears. Whether the forward autocast is set to the default bf16 or manually switched to fp32, the training loss remains inconsistent between two experiments. If we freeze all parameters of the VLM, the training loss becomes consistent across two runs. The reason may lie in the backpropagation of DIT's loss to the shallow layers of the VLM, where the gradients are too small, leading to some kind of numerical instability? This issue currently has no solution.

did you reproduce the success rate in paper 51.3% ? I wonder how much you could achieve. I also plan to use 4 GPUs, bs 80, to reproduce it.

**Comment by fx-hit** [comment]
> > Thank you for your reply. We have resolved this issue but encountered a new non-reproducible problem.
> > We intercepted a small-scale dataset from the tfds data reading pipeline and stored it locally, then handed it over to PyTorch for reading. In other words, the code has been completely switched to PyTorch and no longer involves tfds data reading.
> > We have confirmed that all random seeds are fixed, and the data reading order is strictly controlled to be identical. However, when running two experiments with the same configuration, the training loss still varies. This also affects the final success rate evaluation, causing a discrepancy of 5%-10% when evaluating the success rate on window X.
> > We found that if we enable autocast for the forward pass in fp16, the training loss becomes consistent across two runs—but the model produces NaN after a few hundred iterations, requiring GradScaler. However, once GradScaler is added, the loss inconsistency reappears. Whether the forward autocast is set to the default bf16 or manually switched to fp32, the training loss remains inconsistent between two experiments. If we freeze all parameters of the VLM, the training loss becomes consistent across two runs. The reason may lie in the backpropagation of DIT's loss to the shallow layers of the VLM, where the gradients are too small, leading to some kind of numerical instability? This issue currently has no solution.
> 
> did you reproduce the success rate in paper 51.3% ? I wonder how much you could achieve. I also plan to use 4 GPUs to reproduce it.

We used the officially released CogACT-Base model and tested it on the Windows X robotic arm in the Simpler simulation environment, achieving a success rate of 0.48.

**Comment by fx-hit** [comment]
> > > > Thank you for your reply. We have resolved this issue but encountered a new non-reproducible problem.
> > > > We intercepted a small-scale dataset from the tfds data reading pipeline and stored it locally, then handed it over to PyTorch for reading. In other words, the code has been completely switched to PyTorch and no longer involves tfds data reading.
> > > > We have confirmed that all random seeds are fixed, and the data reading order is strictly controlled to be identical. However, when running two experiments with the same configuration, the training loss still varies. This also affects the final success rate evaluation, causing a discrepancy of 5%-10% when evaluating the success rate on window X.
> > > > We found that if we enable autocast for the forward pass in fp16, the training loss becomes consistent across two runs—but the model produces NaN after a few hundred iterations, requiring GradScaler. However, once GradScaler is added, the loss inconsistency reappears. Whether the forward autocast is set to the default bf16 or manually switched to fp32, the training loss remains inconsistent between two experiments. If we freeze all parameters of the VLM, the training loss becomes consistent across two runs. The reason may lie in the backpropagation of DIT's loss to the shallow layers of the VLM, where the gradients are too small, leading to some kind of numerical instability? This issue currently has no solution.
> > > 
> > > 
> > > did you reproduce the success rate in paper 51.3% ? I wonder how much you could achieve. I also plan to use 4 GPUs to reproduce it.
> > 
> > 
> > We used the officially released CogACT-Base model and tested it on the Windows X robotic arm in the Simpler simulation environment, achieving a success rate of 47.9%.
> 
> Thank you, I test on simpler google robot Open Top Drawer and place apple, seems only 2% different from paper.
> 
> VM	variant
> Open Top Drawer and place apple	Open Top Drawer and place apple
> 50.00%	45.19%
> I noticed that you mention training loss. Did you train from scratch or finetune it? How large the training loss will be ?

We started training based on the OpenVLA checkpoint, using the Bridge V2 dataset. The training loss decreased from around 1 to below 0.1 over several hundred iterations. After several thousand iterations, the checkpoint achieved a high success rate in the simulated evaluation task of Window X.
